pep-0684.rst

```diff
@@ -0,0 +1,640 @@
+PEP: 684
+Title: A Per-Interpreter GIL
+Author: Eric Snow <ericsnowcurrently@gmail.com>
+Discussions-To: python-dev@python.org
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 08-Mar-2022
+Python-Version: 3.11
+Post-History: 08-Mar-2022
+Resolution:
+
+.. XXX Split out an informational PEP with all the relevant info,
+   based on the "Consolidating Runtime Global State" section?
+
+Abstract
+========
+
+Since Python 1.5 (1997), CPython users can run multiple interpreters
+in the same process.  However, they have always shared a significant
+amount of global state.  This is a source of bugs, with a growing
+impact as more and more people use the feature.  Furthermore,
+sufficient isolation would facilitate true multi-core parallelism,
+where interpreters no longer share the GIL.  The changes outlined in
+this proposal will result in that level of interpreter isolation.
+
+
+High-Level Summary
+==================
+
+As a high level, this proposal changes CPython in the following ways:
+
+* stops sharing the GIL between interpreters, given sufficient isolation
+* adds several new interpreter config options for isolation settings
+* adds public C-API for fine-grained control when creating interpreters
+* keeps incompatible extensions from causing problems
+
+The GIL
+-------
+
+The GIL protects concurrent access of most of CPython's runtime state.
+So all that GIL-protected global state must move to each interpreter
+before the GIL can.
+
+(In a handful of cases, other mechanisms can be used to ensure
+thread-safe sharing instead, such as locks or "immortal" objects.)
+
+CPython Runtime State
+---------------------
+
+Properly isolating interpreters requires that most of CPython's
+runtime state be stored in the ``PyInterpreterState`` struct.  Currently
+only a portion of it is.  The rest is found either in global variables
+or in ``_PyRuntimeState``.  Most of that will have to be moved.
+
+This directly coincides with an ongoing effort of many years to greatly
+reduce internal use of C global variables and consolidate the runtime
+state into ``_PyRuntimeState`` and ``PyInterpreterState``.
+(See `Consolidating Runtime Global State`_ below.)  That project has
+`significant merit on its own <Benefits to Consolidation_>`_
+and has faced little controversy.  So, while a per-interpreter GIL
+relies on the completion of that effort, that project should not be
+considered a part of this proposal, only a dependency.
+
+Other Isolation Considerations
+------------------------------
+
+CPython's interpreters must be strictly isolated from each other, with
+few exceptions.  To a large extent they already are.  Each interpreter
```

Comment:
```suggestion
few exceptions.  To a large extent, they already are.  Each interpreter
```

Slight pause improves readability/flow and reduces choppiness somewhat

Reply:
In this case I prefer it without the comma.